x/mongo/driver/topology/server.go

```diff
@@ -662,8 +666,7 @@ func (s *Server) createConnection() (*connection, error) {
 		// Override any monitors specified in options with nil to avoid monitoring heartbeats.
 		WithMonitor(func(*event.CommandMonitor) *event.CommandMonitor { return nil }),
 		withPoolMonitor(func(*event.PoolMonitor) *event.PoolMonitor { return nil }),
-	}
-	opts = append(s.cfg.connectionOpts, opts...)
```

Comment:
To check my understanding: this may modify `cfg.connectionOpts`, but may not.

If the backing array is too small and is reallocated, this will not modify `cfg.connectionOpts`.

That might explain why this is not a consistent problem.

Reply:
That sounds correct to me. We now want to reallocate every time with a copy so that we're never modifying `s.cfg.connectionOpts`.

It seems like in <= 1.5.4, the `s.cfg.connectionOpts` was always too small, and `s.cfg.connectionOpts` was reallocated. So, the data race wasn't an issue. But, in 1.6.0, `s.cfg.connectionOpts` is usually large enough to do this appension in-place, which directly modifies `s.cfg.connectionOpts` and makes the data race a problem.

Reply:
> See an example here: play.golang.org/p/BZeL7fwd-pL

Thank you for the example, that is helpful.

> It seems like in <= 1.5.4, the s.cfg.connectionOpts was always too small, and s.cfg.connectionOpts was reallocated. So, the data race wasn't an issue. But, in 1.6.0, s.cfg.connectionOpts is usually large enough to do this appension in-place, which directly modifies s.cfg.connectionOpts and makes the data race a problem.

That sounds like a reasonable hypothesis to me.

Reply:
I've found the change that surfaced this bug. In v1.6.0, a new option is appended in `topology.newPool()`:
```go
pool.opts = append(pool.opts, withGenerationNumberFn(func(_ generationNumberFn) generationNumberFn { return pool.getGenerationForNewConnection }))
```

Check out the diff:
```
git diff v1.5.4 v1.6.0 x/mongo/driver/topology/pool.go
```
Go to line 54 in the diff (type `n54` if you're using `less` to view diffs) to see the newly appended `ConnectionOption` (the new option is `withGenerationNumberFn`). Appending the new option opens up the view onto the underlying array by 1 element, resulting in created connections being able to see the subsequently appended option here:
https://github.com/mongodb/mongo-go-driver/blob/v1.6.0/x/mongo/driver/topology/server.go#L654

Note that the `connectTimeout` set from `Server.createConnection()` was also leaking to `pool.opts` in v1.5.4, but that doesn't have the same effect as setting `readTimeout`. Also note that the bug appears to be triggered by calling `ClientOptions.SetConnectTimeout()`; not adding that option prevents the bug from surfacing.

Reply:
Thanks for writing this up and investigating so deeply 🚀  Sounds exactly right to me.